File: examples/k8s-sandbox/kafka-throughput/producer/producer.py
import json
import logging
import os
import random
import time
from math import floor
from functools import lru_cache

from kafka import KafkaProducer  # type: ignore[import-not-found]
from kafka.errors import (  # type: ignore[import-not-found]
    KafkaTimeoutError,
    NoBrokersAvailable,
)

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)

# Configuration from environment variables
KAFKA_BROKER = os.environ.get(
    "KAFKA_BROKER", "kafka-service.kafka.svc.cluster.local:9092"
)
KAFKA_TOPIC = os.environ.get("KAFKA_TOPIC", "throughput-topic")
INITIAL_MESSAGE_RATE = int(os.environ.get("PRODUCER_TARGET_RATE", "1000"))
MESSAGE_SIZE_BYTES = int(os.environ.get("MESSAGE_SIZE_BYTES", "1024"))
PRODUCER_CLIENT_ID = os.environ.get("PRODUCER_CLIENT_ID", "kafka-throughput-producer")
MAX_RETRIES = int(os.environ.get("MAX_RETRIES", "30"))
INITIAL_RETRY_DELAY_S = int(os.environ.get("INITIAL_RETRY_DELAY_S", "5"))
STATUS_DIR = os.environ.get("STATUS_DIR", "/tmp/status")
PRODUCER_STATS_FILE = os.path.join(STATUS_DIR, "producer_stats.txt")
LATENCY_FILE = os.path.join(STATUS_DIR, "e2e_latency_ms.txt")
KMINION_PRODUCER_ACTUAL_RATE_FILE = os.path.join(
    STATUS_DIR, "producer_actual_rate_value.txt"
)

# Adaptive Load Parameters
ADAPTIVE_CHECK_INTERVAL_S = int(os.environ.get("ADAPTIVE_CHECK_INTERVAL_S", "10"))
TARGET_LATENCY_MS = float(os.environ.get("TARGET_LATENCY_MS", "5.0"))
MAX_MESSAGE_RATE = int(os.environ.get("MAX_MESSAGE_RATE", "50000"))
RATE_INCREASE_FACTOR = float(os.environ.get("RATE_INCREASE_FACTOR", "1.03"))
ACTUAL_VS_TARGET_THRESHOLD_PERCENT = float(
    os.environ.get("ACTUAL_VS_TARGET_THRESHOLD_PERCENT", "0.95")
)


# Memoized helper for creating padding string
@lru_cache(maxsize=128)
def _get_padding_string(length: int) -> str:
    if length <= 0:
        return ""
    return "x" * length


def create_message(sequence_number: int, size_bytes: int) -> bytes:
    """Creates a JSON message with a sequence number and padding."""
    payload = {
        "sequence": sequence_number,
        "timestamp": time.time(),
        "data": "",  # Placeholder for padding
    }
    # First serialization to determine the exact size of other fields
    base_payload_str = json.dumps(payload, separators=(",", ":"))
    current_encoded_size = len(base_payload_str.encode("utf-8"))

    padding_needed = size_bytes - current_encoded_size

    # Get padding string using the memoized helper
    payload["data"] = _get_padding_string(padding_needed)

    # Second serialization with the actual padding
    message_str = json.dumps(payload, separators=(",", ":"))
    return message_str.encode("utf-8")


def send_message(
    producer: KafkaProducer,
    topic: str,
    sequence_number: int,
    message_size_bytes: int,
) -> bool:
    """
    Creates a message, determines partition, sends it, and handles errors.
    Returns True if successful, False otherwise.
    """
    message_bytes = create_message(sequence_number, message_size_bytes)

    # Partition key logic (Benford's Law like distribution)
    # This attempts to create a non-uniform distribution of messages across partitions
    # for more realistic load.
    lognorm_value = random.lognormvariate(5, 3)
    random_int = floor(lognorm_value)
    partition_key = int(str(random_int)[0])

    try:
        producer.send(topic, value=message_bytes, partition=partition_key)
        logging.debug(f"Sent msg {sequence_number} to partition '{partition_key}'")
        return True
    except KafkaTimeoutError:
        logging.warning("Timeout sending message. Kafka busy or unavailable?")
        return False
    except Exception as e:
        logging.error(f"Error sending message: {e}")
        return False


def write_producer_stats(target_rate: int, size: int) -> None:
    """Writes producer target rate and message size to a file."""
    stats_str = f"target_rate={target_rate},size={size}"
    try:
        os.makedirs(STATUS_DIR, exist_ok=True)
        with open(PRODUCER_STATS_FILE, "w") as f:
            f.write(stats_str)
        logging.debug(f"Producer stats written: {stats_str}")
    except Exception as e:
        logging.error(f"Error writing producer stats to {PRODUCER_STATS_FILE}: {e}")


def _read_metric_from_file(file_path: str, metric_name: str) -> float | None:
    """Helper function to read a metric value from a given file."""
    try:
        if os.path.exists(file_path):
            with open(file_path) as f:
                content = f.read().strip()
                if content and content != "N/A":
                    logging.debug(f"{metric_name} from {file_path}: '{content}'")
                    return float(content)
                elif content == "N/A":
                    logging.debug(f"{metric_name} file {file_path} reported N/A.")
                    return None  # Treat N/A as no data
        else:
            logging.warning(f"{metric_name} file {file_path} not found.")
    except ValueError:
        logging.warning(f"Could not parse float from {metric_name} file: {file_path}")
    except Exception as e:
        logging.error(f"Error reading {metric_name} file {file_path}: {e}")
    return None


def adjust_message_rate_autonomously(
    actual_rate: float | None,
    target_rate: int,
    actual_latency_ms: float | None,
    target_latency_ms: float,
    max_rate_cap: int,
    increase_factor: float,
    actual_vs_target_ratio_threshold: float,
) -> int:
    """Producer will adjust rate upwards if conditions are good."""

    if actual_rate is None:
        logging.info(
            "No KMinion-observed actual rate data available. "
            f"Holding current target rate: {target_rate}/s."
        )
        return target_rate

    if actual_latency_ms is None:
        logging.info(f"No latency data. Holding current target rate: {target_rate}/s.")
        return target_rate

    if actual_latency_ms >= target_latency_ms:
        logging.info(
            f"Latency ({actual_latency_ms:.2f}ms) >= threshold ({target_latency_ms}ms). "
            f"Holding current target rate: {target_rate}/s."
        )
        return target_rate

    if target_rate >= max_rate_cap:
        logging.info(
            f"Latency low ({actual_latency_ms:.2f}ms), but already at max rate cap "
            f"({target_rate}/s). Holding."
        )
        return target_rate

    if actual_rate < (target_rate * actual_vs_target_ratio_threshold):
        logging.info(
            f"Latency low ({actual_latency_ms:.2f}ms < {target_latency_ms}ms), "
            f"but local actual rate ({actual_rate:.2f}/s) is below threshold "
            f"({actual_vs_target_ratio_threshold * 100:.0f}%) of current target "
            f"({target_rate}/s). Holding current target rate."
        )
        return target_rate

    calculated_new_rate = target_rate * increase_factor
    new_rate = min(int(calculated_new_rate), max_rate_cap)

    if new_rate > target_rate:
        logging.info(
            "Conditions good. Increasing target rate from "
            f"{target_rate} to {new_rate}/s."
        )
        return new_rate
    else:
        logging.info(
            f"Conditions good, but proposed new rate {new_rate}/s "
            f"not > current {target_rate}/s. Holding."
        )
        return target_rate


def main() -> None:
    logging.info("--- Kafka Throughput Demo Producer (v2 - Autonomous) --- ")
    logging.info(f"Target Kafka Broker: {KAFKA_BROKER}")
    logging.info(f"Target Kafka Topic: {KAFKA_TOPIC}")

    current_message_rate = INITIAL_MESSAGE_RATE
    logging.info(
        f"Initial Message Rate: {current_message_rate}/s (Max Cap for "
        f"Autonomous Adjustments: {MAX_MESSAGE_RATE}/s)"
    )
    logging.info(f"Message Size: {MESSAGE_SIZE_BYTES} bytes")
    logging.info(f"Client ID: {PRODUCER_CLIENT_ID}")
    logging.info(
        f"Autonomous Adaptive Logic: Check Interval={ADAPTIVE_CHECK_INTERVAL_S}s, "
        f"Target Latency=<{TARGET_LATENCY_MS}ms, "
        f"Increase Factor=x{RATE_INCREASE_FACTOR}, "
        f"Min Actual/Target Ratio={ACTUAL_VS_TARGET_THRESHOLD_PERCENT * 100:.0f}%"
    )

    producer = None
    retries = 0
    while retries < MAX_RETRIES:
        try:
            producer = KafkaProducer(
                bootstrap_servers=KAFKA_BROKER,
                client_id=PRODUCER_CLIENT_ID,
                acks=1,
                retries=3,
                linger_ms=50,
                batch_size=16384 * 16,
            )
            logging.info("KafkaProducer initialized successfully.")
            break
        except NoBrokersAvailable as e:
            retries += 1
            logging.warning(
                f"Attempt {retries}/{MAX_RETRIES}: Could not connect to Kafka "
                f"broker at {KAFKA_BROKER}. Error: {e}"
            )
            if retries >= MAX_RETRIES:
                logging.error("Maximum connection attempts reached. Exiting.")
                exit(1)
            delay = INITIAL_RETRY_DELAY_S * (2 ** (retries - 1))
            logging.info(f"Retrying in {delay} seconds...")
            time.sleep(delay)
        except Exception as e:
            logging.error(
                f"An unexpected error during KafkaProducer initialization: {e}"
            )
            exit(1)

    if not producer:
        logging.error("Failed to initialize Kafka producer after multiple retries.")
        exit(1)

    logging.info("Writing initial producer stats (target_rate, size)...")
    write_producer_stats(target_rate=current_message_rate, size=MESSAGE_SIZE_BYTES)

    sequence_number = 0
    messages_sent_this_interval = 0  # For calculating local actual rate over 1 second
    interval_start_time = time.time()
    last_autonomous_adjustment_check_time = time.time()

    try:
        while True:
            now = time.time()

            # 1. Autonomous Adaptive Rate Adjustment (periodically)
            if now - last_autonomous_adjustment_check_time >= ADAPTIVE_CHECK_INTERVAL_S:
                last_autonomous_adjustment_check_time = now

                current_e2e_latency_ms = _read_metric_from_file(
                    LATENCY_FILE, "E2E latency"
                )

                # Fetch KMinion's observed producer rate for decision making
                kminion_observed_producer_rate = _read_metric_from_file(
                    KMINION_PRODUCER_ACTUAL_RATE_FILE,
                    "KMinion observed producer rate for decision",
                )

                adjusted_target_rate = adjust_message_rate_autonomously(
                    actual_rate=kminion_observed_producer_rate,
                    target_rate=current_message_rate,
                    actual_latency_ms=current_e2e_latency_ms,
                    target_latency_ms=TARGET_LATENCY_MS,
                    max_rate_cap=MAX_MESSAGE_RATE,
                    increase_factor=RATE_INCREASE_FACTOR,
                    actual_vs_target_ratio_threshold=ACTUAL_VS_TARGET_THRESHOLD_PERCENT,
                )

                if adjusted_target_rate != current_message_rate:
                    logging.info(
                        "Autonomous adjustment: Target rate changed from "
                        f"{current_message_rate} to {adjusted_target_rate}/s."
                    )
                    current_message_rate = adjusted_target_rate
                    # Write new autonomously adjusted target to stats file
                    write_producer_stats(
                        target_rate=current_message_rate, size=MESSAGE_SIZE_BYTES
                    )

            # 2. Message Sending Loop (to meet current_message_rate over 1 second)
            # This loop aims to send `current_message_rate` messages
            # within each 1-second wall-clock interval.
            if messages_sent_this_interval < current_message_rate:
                if send_message(
                    producer, KAFKA_TOPIC, sequence_number, MESSAGE_SIZE_BYTES
                ):
                    sequence_number += 1
                    messages_sent_this_interval += 1
                else:
                    # Prevents a very tight loop if Kafka is temporarily unresponsive.
                    time.sleep(0.1)

            # 3. Stats Reporting & Loop Timing (every 1 second)
            # Also ensures the loop roughly aligns with 1-second iterations.
            if now - interval_start_time >= 1.0:
                elapsed_interval_time = now - interval_start_time
                logging.info(
                    f"Interval: {elapsed_interval_time:.2f}s, "
                    f"Current Target Rate: {current_message_rate}/s, "
                    f"Sent this interval: {messages_sent_this_interval}"
                )

                write_producer_stats(
                    target_rate=current_message_rate, size=MESSAGE_SIZE_BYTES
                )

                # Reset for next 1-second interval
                messages_sent_this_interval = 0
                interval_start_time = now  # Reset interval timer to current time
            else:
                # If we've already sent all messages for this target rate in the
                # current <1s window, sleep briefly to yield CPU and avoid busy-waiting.
                if messages_sent_this_interval >= current_message_rate:
                    # Calculate remaining time in 1-second interval and sleep for that
                    # duration, or a small fixed amount if the calculation is tricky.
                    # This helps maintain the 1-second rhythm more accurately.
                    remaining_time_in_interval = 1.0 - (now - interval_start_time)
                    if (
                        remaining_time_in_interval > 0.001
                    ):  # Only sleep if meaningful time left
                        time.sleep(
                            min(remaining_time_in_interval, 0.1)
                        )  # Sleep up to 0.1s or remaining time
                    else:
                        time.sleep(
                            0.001
                        )  # Minimal sleep if very close to end of interval or overshot

    except KeyboardInterrupt:
        logging.info("Producer interrupted. Shutting down...")
    except Exception as e:
        logging.error(f"An unexpected error occurred in main loop: {e}", exc_info=True)
    finally:
        if producer:
            logging.info("Flushing remaining messages...")
            producer.flush(timeout=10)  # Adding timeout to flush
            logging.info("Closing Kafka producer...")
            producer.close(timeout=10)  # Adding timeout to close
            logging.info("Producer closed.")
            # Write final stats (target_rate, size) on clean exit
            write_producer_stats(
                target_rate=current_message_rate, size=MESSAGE_SIZE_BYTES
            )


if __name__ == "__main__":
    main()

File: examples/k8s-sandbox/kafka-throughput/producer/producer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"os/signal"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"

	"github.com/IBM/sarama"
	"github.com/google/uuid"
)

// Configuration from environment variables with defaults
var (
	kafkaBroker                    = getEnv("KAFKA_BROKER", "k8s-sandbox:9092")
	kafkaTopic                     = getEnv("KAFKA_TOPIC", "throughput-topic")
	producerClientID               = getEnv("PRODUCER_CLIENT_ID", "kafka-throughput-producer")
	statusDir                      = getEnv("STATUS_DIR", "/tmp/status")
	producerStatsFile              = filepath.Join(statusDir, "producer_stats.txt")
	latencyFile                    = filepath.Join(statusDir, "e2e_latency_ms.txt")
	kminionProducerActualRateFile  = filepath.Join(statusDir, "producer_actual_rate_value.txt")
	initialMessageRate, _          = strconv.Atoi(getEnv("MESSAGE_RATE_PER_SECOND", "1000"))
	messageSizeBytes, _            = strconv.Atoi(getEnv("MESSAGE_SIZE_BYTES", "1024"))
	maxRetries, _                  = strconv.Atoi(getEnv("MAX_RETRIES", "30"))
	initialRetryDelayS, _          = strconv.Atoi(getEnv("INITIAL_RETRY_DELAY_S", "5"))
	targetLatencyMs, _             = strconv.ParseFloat(getEnv("TARGET_LATENCY_MS", "5.0"), 64)
	targetRateMargin, _            = strconv.ParseFloat(getEnv("TARGET_RATE_MARGIN", "0.80"), 64)
	targetRateMult, _              = strconv.ParseFloat(getEnv("TARGET_RATE_MULT", "1.03"), 64)
	maxMessageRate, _              = strconv.Atoi(getEnv("MAX_MESSAGE_RATE", "50000"))
)

var partitionSkew float64

// init function for robust parsing and validation of partitionSkew
func init() {
	skewStr := getEnv("PARTITION_SKEW", "0.0")
	var errParse error
	partitionSkew, errParse = strconv.ParseFloat(skewStr, 64)
	if errParse != nil {
		log.Fatalf("Invalid PARTITION_SKEW value: %s. Error: %v", skewStr, errParse)
	}
	if partitionSkew < 0.0 || partitionSkew > 1.0 {
		log.Fatalf("PARTITION_SKEW must be between 0.0 and 1.0, got: %.2f", partitionSkew)
	}
	// No need to log here, main will log it if needed or it's used directly
}

// Message represents our Kafka message payload
type Message struct {
	Sequence   int64     `json:"sequence"`
	Timestamp  float64   `json:"timestamp"`
	Data       string    `json:"data"`
}

// ProducerStats tracks producer statistics
type ProducerStats struct {
	mu            sync.Mutex
	targetRate    int
	messageSize   int
	lastWriteTime time.Time
}

// Global stats instance
var stats = &ProducerStats{
	targetRate:  initialMessageRate,
	messageSize: messageSizeBytes,
}

// BatchConfig holds batching configuration
type BatchConfig struct {
	MaxBatchSize  int
	MaxBatchDelay time.Duration
}

// MessageBatch represents a batch of messages to be sent
type MessageBatch struct {
	messages []*sarama.ProducerMessage
	size     int
}

// Global batch config
const (
	batchesPerSecond = 20 // 50ms between batches
)

var batchConfig = BatchConfig{
	MaxBatchSize:  5000, // Max messages in an app-level batch. Sarama also has its own flush limits.
	MaxBatchDelay: time.Second / time.Duration(batchesPerSecond),
}

var effectiveMaxTokensForBucket float64 // New global for token bucket capacity

// TokenBucket implements a simple token bucket rate limiter
type TokenBucket struct {
	mu            sync.Mutex
	tokens        float64
	rate          float64    // tokens per second
	lastRefill    time.Time
	maxTokens     float64
}

func newTokenBucket(rate float64, configuredMaxTokens float64) *TokenBucket {
	now := time.Now()
	initialTokens := configuredMaxTokens // Start full
	return &TokenBucket{
		tokens:     initialTokens,
		rate:       rate,
		lastRefill: now,
		maxTokens:  configuredMaxTokens,
	}
}

func (tb *TokenBucket) refill() {
	now := time.Now()
	elapsed := now.Sub(tb.lastRefill).Seconds()
	newTokens := elapsed * tb.rate
	tb.tokens = math.Min(tb.maxTokens, tb.tokens + newTokens)
	tb.lastRefill = now
	log.Printf("TokenBucket.refill: refilled %.2f tokens, total now %.2f", newTokens, tb.tokens)
}

func (tb *TokenBucket) take(n float64) bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	log.Printf("TokenBucket.take: attempting to take %.2f tokens", n)
	tb.refill()
	log.Printf("TokenBucket.take: after refill, have %.2f tokens", tb.tokens)
	if tb.tokens >= n {
		tb.tokens -= n
		log.Printf("TokenBucket.take: took %.2f tokens, %.2f remaining", n, tb.tokens)
		return true
	}
	log.Printf("TokenBucket.take: not enough tokens (need %.2f, have %.2f)", n, tb.tokens)
	return false
}

func (tb *TokenBucket) setRate(rate float64) {
	tb.mu.Lock()
	defer tb.mu.Unlock()
	tb.rate = rate
}

// getEnv gets an environment variable or returns a default value
func getEnv(key, defaultValue string) string {
	if value, exists := os.LookupEnv(key); exists {
		return value
	}
	return defaultValue
}

// writeProducerStats writes the current producer stats to file
func (p *ProducerStats) writeStats() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	// Only write if more than 1 second has passed since last write
	if time.Since(p.lastWriteTime) < time.Second {
		return nil
	}

	statsStr := fmt.Sprintf("target_rate=%d,size=%d", p.targetRate, p.messageSize)
	if err := os.MkdirAll(statusDir, 0755); err != nil {
		return fmt.Errorf("failed to create status directory: %w", err)
	}

	if err := os.WriteFile(producerStatsFile, []byte(statsStr), 0644); err != nil {
		return fmt.Errorf("failed to write producer stats: %w", err)
	}

	p.lastWriteTime = time.Now()
	return nil
}

// readMetricFromFile reads a metric value from a file
func readMetricFromFile(filePath string) (float64, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return 0, err
	}

	value, err := strconv.ParseFloat(string(data), 64)
	if err != nil {
		return 0, fmt.Errorf("failed to parse metric value: %w", err)
	}

	return value, nil
}

// createMessage creates a message with the given sequence number and size
func createMessage(sequence int64, sizeBytes int) ([]byte, error) {
	// Calculate padding size (accounting for JSON overhead)
	baseMsg := Message{
		Sequence:  sequence,
		Timestamp: float64(time.Now().UnixNano()) / 1e9,
		Data:      "",
	}

	baseJSON, err := json.Marshal(baseMsg)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal base message: %w", err)
	}

	paddingSize := sizeBytes - len(baseJSON)
	if paddingSize < 0 {
		paddingSize = 0
	}

	// Create the final message with padding
	msg := Message{
		Sequence:  sequence,
		Timestamp: float64(time.Now().UnixNano()) / 1e9,
		Data:      string(make([]byte, paddingSize)),
	}

	return json.Marshal(msg)
}

// adjustMessageRate adjusts the message rate based on current conditions
func adjustMessageRate(currentRate int, actualRate float64, actualLatency float64) int {
	// If we don't have metrics yet, hold current rate
	if actualRate == 0 || actualLatency == 0 {
		return currentRate
	}

	// If latency is too high, hold current rate
	if actualLatency >= targetLatencyMs {
		return currentRate
	}

	// If actual rate is below target rate margin, hold current rate
	if actualRate < float64(currentRate)*targetRateMargin {
		return currentRate
	}

	// Increase rate by configured multiplier
	newRate := int(float64(currentRate) * targetRateMult)

	// Cap at configured max message rate
	if newRate > maxMessageRate {
		newRate = maxMessageRate
	}

	return newRate
}

// createMessageBatch creates a batch of messages
func createMessageBatch(startSeq int64, count int, sizeBytes int) (*MessageBatch, error) {
	log.Printf("Creating message batch: startSeq=%d, count=%d, sizeBytes=%d, skew_factor=%.2f", startSeq, count, sizeBytes, partitionSkew)
	batch := &MessageBatch{
		messages: make([]*sarama.ProducerMessage, 0, count),
	}

	for i := 0; i < count; i++ {
		msgBytes, err := createMessage(startSeq+int64(i), sizeBytes)
		if err != nil {
			log.Printf("Error creating message %d: %v", i, err)
			return nil, err
		}

		var keyEncoder sarama.Encoder
		var producerMsg *sarama.ProducerMessage

		if rand.Float64() < partitionSkew {
			keyEncoder = sarama.StringEncoder("skew_key_0")
			producerMsg = &sarama.ProducerMessage{
				Topic:     kafkaTopic,
				Value:     sarama.ByteEncoder(msgBytes),
				Key:       keyEncoder,
				Partition: 0,
			}
		} else {
			keyEncoder = sarama.StringEncoder(uuid.New().String())
			producerMsg = &sarama.ProducerMessage{
				Topic: kafkaTopic,
				Value: sarama.ByteEncoder(msgBytes),
				Key:   keyEncoder,
			}
		}

		batch.messages = append(batch.messages, producerMsg)
		batch.size++
	}
	log.Printf("Successfully created batch of %d messages", batch.size)
	return batch, nil
}

// sendBatch sends a batch of messages
func sendBatch(producer sarama.AsyncProducer, batch *MessageBatch) {
	log.Printf("sendBatch: Sending %d messages to AsyncProducer.Input()", batch.size)
	for i, msg := range batch.messages {
		select {
		case producer.Input() <- msg:
		case <-time.After(1 * time.Second): // Timeout to prevent deadlock if Input() is full and not draining
			log.Printf("sendBatch: Timeout enqueuing message %d/%d to producer.Input(). Producer buffer might be full.", i+1, batch.size)
			// Optionally, we could stop trying to send the rest of the batch here
			return
		}
	}
	log.Printf("sendBatch: All %d messages enqueued to AsyncProducer.Input()", batch.size)
}

func main() {
	log.SetFlags(log.LstdFlags | log.Lmicroseconds)
	log.Printf("--- Kafka Throughput Demo Producer (Go - Async) ---")
	log.Printf("Target Kafka Broker: %s", kafkaBroker)
	log.Printf("Target Kafka Topic: %s", kafkaTopic)
	log.Printf("Initial Message Rate: %d/s", initialMessageRate)
	log.Printf("Message Size: %d bytes", messageSizeBytes)
	log.Printf("Client ID: %s", producerClientID)
	log.Printf("Max Message Rate: %d/s", maxMessageRate)
	log.Printf("Target Rate Margin: %.2f", targetRateMargin)
	log.Printf("Target Rate Multiplier: %.2f", targetRateMult)

	// partitionSkew is initialized in the init() function
	log.Printf("Partition Skew Factor: %.2f", partitionSkew)

	// Calculate effectiveMaxTokensForBucket after all env vars are parsed
	effectiveMaxTokensForBucket = float64(maxMessageRate / batchesPerSecond)
	if effectiveMaxTokensForBucket < 100 { // Ensure a minimum reasonable capacity
		effectiveMaxTokensForBucket = 100
	}
	log.Printf("Token Bucket Max Tokens (Capacity): %.2f", effectiveMaxTokensForBucket)

	config := sarama.NewConfig()
	config.Producer.RequiredAcks = sarama.WaitForLocal
	config.Producer.Retry.Max = 3
	config.Producer.Flush.Frequency = 5 * time.Millisecond
	config.Producer.Flush.Bytes = 5 * 1024 * 1024
	config.Producer.Flush.Messages = 5000
	config.Producer.Compression = sarama.CompressionSnappy
	config.Producer.MaxMessageBytes = 1000000
	config.Net.MaxOpenRequests = 20
	config.Producer.Return.Successes = true // Important for AsyncProducer
	config.Producer.Return.Errors = true   // Important for AsyncProducer
	config.ClientID = producerClientID
	config.ChannelBufferSize = 1024

	var producer sarama.AsyncProducer // Changed to AsyncProducer
	var err error
	for retries := 0; retries < maxRetries; retries++ {
		producer, err = sarama.NewAsyncProducer([]string{kafkaBroker}, config)
		if err == nil {
			log.Printf("AsyncProducer initialized successfully.")
			break
		}
		delay := time.Duration(initialRetryDelayS*(1<<retries)) * time.Second
		log.Printf("Failed to create AsyncProducer (attempt %d/%d): %v. Retrying in %v...",
			retries+1, maxRetries, err, delay)
		time.Sleep(delay)
	}
	if err != nil {
		log.Fatalf("Failed to create AsyncProducer after %d retries: %v", maxRetries, err)
	}
	defer func() {
		log.Printf("Attempting to close AsyncProducer...")
		if err := producer.Close(); err != nil {
			log.Printf("Error closing producer: %v", err)
		}
		log.Printf("AsyncProducer closed.")
	}()

	// Goroutine to handle successful deliveries
	go func() {
		for succ := range producer.Successes() {
			_ = succ // Avoid unused variable error if logging is commented out
		}
		log.Printf("Producer Successes channel closed.")
	}()

	// Goroutine to handle delivery errors
	go func() {
		for err := range producer.Errors() {
			log.Printf("Failed to send message: %v", err)
			// Here you could implement logic to retry, refund tokens, etc.
			// For now, just logging. If a message fails, tokens are still spent.
		}
		log.Printf("Producer Errors channel closed.")
	}()

	// Initialize stats
	if err := stats.writeStats(); err != nil {
		log.Printf("Warning: Failed to write initial stats: %v", err)
	}

	// Setup context for graceful shutdown
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// Handle signals for graceful shutdown
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigChan
		log.Printf("Received shutdown signal. Flushing messages...")
		cancel()
	}()

	// Initialize token bucket for rate control
	tokenBucket := newTokenBucket(float64(stats.targetRate), effectiveMaxTokensForBucket)
	log.Printf("Initialized token bucket with rate %.2f/s, capacity: %.2f, current tokens: %.2f",
		float64(stats.targetRate), effectiveMaxTokensForBucket, tokenBucket.tokens)

	// Message sending loop with batching
	var sequence int64
	intervalStart := time.Now()
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()

	// Rate control variables
	targetRatePerBatch := stats.targetRate / batchesPerSecond
	if targetRatePerBatch < 1 {
		targetRatePerBatch = 1
	}
	log.Printf("Initial batch size: %d messages (target rate: %d/s, batches: %d/s)",
		targetRatePerBatch, stats.targetRate, batchesPerSecond)

	rateLimiter := time.NewTicker(batchConfig.MaxBatchDelay)
	defer rateLimiter.Stop()

	log.Printf("Starting main message loop with batch delay: %v", batchConfig.MaxBatchDelay)

	for {
		select {
		case <-ctx.Done():
			log.Printf("Shutting down producer via ctx.Done()...")
			// producer.AsyncClose() // Start the graceful shutdown process for AsyncProducer
			return
		case <-ticker.C:
			log.Printf("Ticker tick received, checking metrics...")
			// Calculate actual rate for this interval - Now using KMinion's reported rate
			elapsed := time.Since(intervalStart).Seconds()
			// internalActualRate := 0.0 // Keep for logging if desired, but not for rate adjustment
			// if elapsed > 0 { // Avoid division by zero if interval is too short
			// 	internalActualRate = float64(atomic.LoadInt64(&messagesSentThisInterval)) / elapsed
			// }

			kminionActualRate, err := readMetricFromFile(kminionProducerActualRateFile)
			if err != nil {
				log.Printf("Warning: Failed to read KMinion actual rate from %s: %v. Rate adjustment might be impaired.", kminionProducerActualRateFile, err)
				// Optionally, decide on a fallback behavior, e.g., hold current rate or use internal rate if available and desired.
				// For now, if kminion rate is unavailable, adjustMessageRate will receive 0 and likely hold the rate.
			}

			// Read current latency
			actualLatency, err := readMetricFromFile(latencyFile)
			if err != nil {
				// Only log if it's not the expected "N/A" or missing file case
				if err.Error() != "failed to parse metric value: strconv.ParseFloat: parsing \"N/A\": invalid syntax" &&
				   !strings.Contains(err.Error(), "no such file or directory") {
					log.Printf("Warning: Failed to read latency: %v", err)
				}
			}

			// Adjust rate based on conditions using KMinion's rate
			newRate := adjustMessageRate(stats.targetRate, kminionActualRate, actualLatency)
			if newRate != stats.targetRate {
				log.Printf("Adjusting rate from %d to %d/s (KMinion Actual: %.2f/s, Latency: %.2fms)", stats.targetRate, newRate, kminionActualRate, actualLatency)
				stats.targetRate = newRate
				tokenBucket.setRate(float64(newRate))
				// Update batch size for new rate
				targetRatePerBatch = newRate / batchesPerSecond
				if targetRatePerBatch < 1 {
					targetRatePerBatch = 1
				}
				if err := stats.writeStats(); err != nil {
					log.Printf("Warning: Failed to write stats: %v", err)
				}
			}

			// Log interval stats
			log.Printf("Interval: %.2fs, Target Rate: %d/s, KMinion Actual Rate: %.2f/s, Batch Size: %d, Tokens: %.1f",
				elapsed, stats.targetRate, kminionActualRate,
				targetRatePerBatch, tokenBucket.tokens)

			// Reset for next interval
			// atomic.StoreInt64(&messagesSentThisInterval, 0) // No longer primary
			intervalStart = time.Now()

		case <-rateLimiter.C:
			log.Printf("Rate limiter tick received, checking tokens...")
			tokensNeeded := float64(targetRatePerBatch)
			currentTokens := tokenBucket.tokens // For logging before take()
			log.Printf("Current tokens: %.2f, needed: %.2f", currentTokens, tokensNeeded)

			if tokenBucket.take(tokensNeeded) {
				log.Printf("Got enough tokens, preparing batch...")
				batchSize := targetRatePerBatch
				if batchSize > batchConfig.MaxBatchSize {
					batchSize = batchConfig.MaxBatchSize
				}
				if batchSize <= 0 { // Ensure batchSize is at least 1 if targetRatePerBatch was 0
					log.Printf("Batch size is %d, skipping batch creation.", batchSize)
					continue
				}
				log.Printf("Calculated batch size: %d", batchSize)

				log.Printf("Creating message batch...")
				currentSequence := atomic.LoadInt64(&sequence)
				batch, err := createMessageBatch(currentSequence, batchSize, stats.messageSize)
				if err != nil {
					log.Printf("Error creating message batch: %v", err)
					tokenBucket.refill() // Refund tokens if batch creation failed
					log.Printf("Refilled token bucket after batch creation error")
					continue
				}
				atomic.AddInt64(&sequence, int64(batch.size)) // Increment sequence by actual batch size
				log.Printf("Successfully created batch, attempting to send...")

				sendBatch(producer, batch) // This is now non-blocking
				// atomic.AddInt64(&messagesSentThisInterval, int64(batch.size)) // No longer primary for rate adjustment
				log.Printf("Enqueued batch of %d messages to AsyncProducer, tokens after take: %.2f",
					batch.size, tokenBucket.tokens)

			} else {
				log.Printf("Not enough tokens (need %.2f, have %.2f), waiting for next tick",
					tokensNeeded, currentTokens)
			}
		}
	}
}
